### PR TITLE
Add BeSameDateAs

### LINKED ISSUE #319: support for comparing date only
Something like DateTime.Now.Should().BeSameDateAs(DateTime.Today)

The closest workaround I can get is

DateTime expectedDate = DateTime.Today;
DateTime.Now.Should().HaveYear(expectedDate.Year)
                .And.HaveMonth(expectedDate.Month)
                .And.HaveDay(expectedDate.Day);

### PR DESCRIPTION
**Pull request by blairconrad**
Fixes #319.

I kept the name suggested in the issue, since you'd not expressed any displeasure.
I'm happy to make any changes you'd like, of course.

### CODE REVIEW

**Comment by blairconrad** [comment]
Oh, I'm targeting `master` instead of `develop` like it says in [CONTRIBUTING.md](https://github.com/dennisdoomen/fluentassertions/blob/master/CONTRIBUTING.md), because I couldn't find `develop`.
I hope that's alright. I'm tempted to submit a PR against CONTRIBUTING.md, as I fear it's out of date…
And I'm happy to update the docs, but thought it premature at this point.

**Comment by dennisdoomen** [comment]
Yeah, you're right. I've recently dropped the `develop` but forgot to update the contribution docs.

File: Tests/FluentAssertions.Shared.Specs/DateTimeAssertionSpecs.cs [hunk 1]
```diff
@@ -792,6 +792,47 @@ public void Should_fail_with_descriptive_message_when_asserting_datetime_has_sec
                 .WithMessage("Expected second 1 because we want to test the failure message, but found 0.");
         }
 
+        [TestMethod]
+        public void Should_succeed_when_asserting_datetime_is_on_same_date()
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Shared.Specs/DateTimeAssertionSpecs.cs]: Please write the specs using the same layout and naming conventions as https://github.com/dennisdoomen/fluentassertions/blob/master/Tests/FluentAssertions.Shared.Specs/BasicEquivalencySpecs.cs#L33

File: Src/Core/Primitives/DateTimeAssertions.cs [hunk 2]
```diff
@@ -435,5 +435,37 @@ public DateTimeRangeAssertions BeLessThan(TimeSpan timeSpan)
...
+        /// </param>
+        public AndConstraint<DateTimeAssertions> BeSameDateAs(DateTime expected, string because = "",
+            params object[] reasonArgs)
+        {
+            bool success = Execute.Assertion
+                .ForCondition(Subject.HasValue)
+                .BecauseOf(because, reasonArgs)
+                .FailWith("Expected a {context:date and time} with date {0}{reason}, but found a <null> DateTime.", expected);
+
+            if (success)
```
Comment (dennisdoomen) [inline Src/Core/Primitives/DateTimeAssertions.cs]: You can avoid the need of a separate `success` variable by chaining assertions like https://github.com/dennisdoomen/fluentassertions/blob/master/Src/Core/Collections/CollectionAssertions.cs#L37

**Comment by dennisdoomen** [comment]
Done reviewing. Would you be willing to take another look?

File: Tests/FluentAssertions.Shared.Specs/DateTimeAssertionSpecs.cs (hunk 1 above)
Comment (blairconrad) [inline Tests/FluentAssertions.Shared.Specs/DateTimeAssertionSpecs.cs]: Happily. Was not sure which convention to follow from `DateTimeAssertionSpecs` and CONTRIBUTING.md referenced a now-extinct file. I'll meet you later for a CONTRIBUTING.md PR.

File: Src/Core/Primitives/DateTimeAssertions.cs (hunk 2 above)
Comment (blairconrad) [inline Src/Core/Primitives/DateTimeAssertions.cs]: Will do!

**Comment by blairconrad** [comment]
Thanks for the comments. I'd be thrilled to take another look.
Oh, while I have you here, do you prefer an additional commit to address issues, or a replacement commit force pushed to the branch (I'd be inclined to go for the latter, but some people like to see the commits match the conversation history, I've found).

**Comment by blairconrad** [comment]
Oh, and assuming I get things wrangled to your satisfaction, what's your policy on cleaning up touched files? I could take a crack at improving consistency and applying the changes you've wanted within those two files. (Generally I'd do this as a second commit in the same PR, but am totally flexible.)
Or I could do nothing. Whatever you like!

**Comment by dennisdoomen** [comment]
Thanks @blairconrad 

I usually amend and force push my rework into the PR since for small PRs I don't care too much about the rework as long as the final solution is okay.
With respect to cleaning up, I sometimes to that myself, but tend to keep them in a separate commit. It's up to you if you want to spend time on that. 

PS. Thanks for the update of the contribution guidelines.

**Comment by dennisdoomen** [comment]
I'll wait for your rework before I tag the 4.2.0 release.

**Comment by blairconrad** [comment]
Oh, I'd definitely clean up in a separate commit.
How's about this:

- I'll rework, amend, and force push my change.
- It will be fabulous, and you'll be very happy.
- Then you can tag the release.
- We'll consider the cleanup/tidy when there's no release pressure. Which still might be tonight. Or tomorrow or something.

I _should_ be able to carve out time for the rework tonight (I'm in UTC-5).

And I'm happy to have updated the contribution guidelines. Everyone wins.

**Comment by dennisdoomen** [comment]
:+1:

**Comment by blairconrad** [comment]
I think I've addressed all your concerns. But names are hard (for me). Please complain about the spec names if they are inadequate.
Oh. I was not sure I understood `Then`, so I snuck in a bonus nullable DateTime spec without consultation. I hope that's okay.
If not, I will happily remove.

**Comment by lukefan6** [comment]
Hi I just wanna say thank you all for making my issue / request resolved.
:+1:  :+1:  :+1:  :100: 
Looking forward to the next release :smile:

**Comment by dennisdoomen** [comment]
It's released: http://www.nuget.org/packages/FluentAssertions/4.2.0

**Comment by blairconrad** [comment]
Thanks, @dennisdoomen.
@lukefan6, you're welcome. Enjoy.

**Pull request merged**